pkg/abcupdater/version.go

```diff
@@ -127,9 +129,53 @@ func (c *CheckVersionParams) setLocalCachedData(data *LocalVersionData) error {
 	return nil
 }
 
-// CheckAppVersion checks if a newer version of an app is available. Any relevant update info will be
+// CheckAppVersion calls CheckAppVersionSync in a go routine. It returns a closure
+// to be run after program logic which will block until a response is returned
+// or a timeout is hit. Results will be printed to stderr.
+func CheckAppVersion(ctx context.Context, params *CheckVersionParams) func() {
+	return asyncFunctionCall(ctx, func() (string, error) {
+		return CheckAppVersionSync(ctx, params)
+	},
+		os.Stderr)
+}
+
+// asyncFunctionCall handles the async part of CheckAppVersion, but accepts
+// a function and writer as an argument to allow for testability.
+func asyncFunctionCall(ctx context.Context, funcToCall func() (string, error), out io.Writer) func() {
```

Comment:
I don't actually think this is safe, because you can't be certain that the writer isn't busy doing other stuff? I think this will spit out the warning in the middle of other output instead of "at shutdown".

Reply:
That is a good question.

The writer is only used in the closure returned by this function, which doesn't get called until shutdown.

It will be called like

```go
results := CheckAppVersion(ctx, params)
defer results()
// rest of program
```
`CheckAppVersion` calls the internal `asyncFunctionCall` in a goroutine, and returns a functions (`results` in my example) which blocks until something comes from the channel.

Since the writing to the writer (`stderr` except for my testing) happens in the defered blocking code, we know nothing else is running (assuming there isn't some other goroutine).

Reply:
Fair. I guess it depends on how much you trust the caller. I think I prefer the architecture where the result is returned as a string to the app, simply because it doesn't require each app to implement a custom `writer` interface.

Reply:
Hmm, I see what you are saying, have `results()` return a string, though note the caller can never provide a `writer`, the exported function hardcodes it to `stderr`. The only reason I accept a writer here is for testability, though I  could do the same by returning a string.

I think I'll move to returning a string anyways, as its cleaner.

Reply:
Yea - the problem with the string approach is that you have to return some kind of "waiter" back to the caller, like a channel or a context.

Reply:
I would keep the blocking function to encapsulate the waiter

```
defer fmt.Fprintf(os.Stderr, "%s", results())
```

Though that really doesn't gain you anything unless you want to print somewhere besides stderr.

Reply:
What if you accepted a function instead?

```go
results := CheckAppVersion(ctx, params, func(s string) {
  cli.Write(s)
})
defer results()
```

Reply:
Isn't that just an `io.Writer` with extra steps?

Reply:
No because it's a function closure, so you can do whatever (including access state and parameters of the parent)

Reply:
Changed